native/src/main/java/io/ballerina/stdlib/data/jsondata/json/JsonTraverse.java

```diff
@@ -0,0 +1,301 @@
+                    }
+                    break;
+                case TypeTags.TUPLE_TAG:
+                    TupleType tupleType = (TupleType) rootArray;
+                    Type restType = tupleType.getRestType();
+                    int expectedTupleTypeCount = tupleType.getTupleTypes().size();
+                    if (expectedTupleTypeCount > array.getLength()) {
+                        throw DiagnosticLog.error(DiagnosticErrorCode.ARRAY_SIZE_MISMATCH);
+                    }
+
+                    for (int i = 0; i < array.getLength(); i++) {
+                        Object jsonMember = array.get(i);
+                        Object nextJsonNode;
+                        if (i < expectedTupleTypeCount) {
+                            nextJsonNode = traverseJson(jsonMember, tupleType.getTupleTypes().get(i));
+                        } else if (restType != null) {
+                            nextJsonNode = traverseJson(jsonMember, restType);
+                        } else {
+                            continue;
+                        }
+                        ((BArray) currentJsonNode).add(i, nextJsonNode);
+                    }
+                    break;
+            }
+            return currentJsonNode;
+        }
+
+        private void traverseArrayMembers(long length, BArray array, Type elementType, Object currentJsonNode) {
+            for (int i = 0; i < length; i++) {
+                ((BArray) currentJsonNode).add(i, traverseJson(array.get(i), elementType));
+            }
+        }
+
+        private void addRestField(Type restFieldType, BString key, Object jsonMember, Object currentJsonNode) {
+            Object nextJsonValue;
+            switch (restFieldType.getTag()) {
+                case TypeTags.ANYDATA_TAG:
+                case TypeTags.JSON_TAG:
+                    ((BMap<BString, Object>) currentJsonNode).put(key, jsonMember);
+                    break;
+                case TypeTags.BOOLEAN_TAG:
+                case TypeTags.INT_TAG:
+                case TypeTags.FLOAT_TAG:
+                case TypeTags.DECIMAL_TAG:
+                case TypeTags.STRING_TAG:
+                    if (checkTypeCompatibility(restFieldType, jsonMember)) {
+                        ((BMap<BString, Object>) currentJsonNode).put(key, jsonMember);
+                    }
+                    break;
+                default:
+                    nextJsonValue = traverseJson(jsonMember, restFieldType);
+                    ((BMap<BString, Object>) currentJsonNode).put(key, nextJsonValue);
+                    break;
+            }
+        }
+
+        private boolean checkTypeCompatibility(Type type, Object json) {
+            return ((json == null && type.getTag() == TypeTags.NULL_TAG)
+                    || (json instanceof BString && type.getTag() == TypeTags.STRING_TAG)
+                    || (json instanceof Long && type.getTag() == TypeTags.INT_TAG)
```

Comment:
Had a offline discussion with @heshanpadmasiri and came to the following conclusion regarding Ballerina `int` datatype.

The Java type `Long` is not the only type that match with Ballerina `int` type.
Ballerina contains several `int` types that match with Java `int` type as well

Reply:
As @heshanpadmasiri suggested, the solution should be something like follows

```
private boolean isInstanceOfBallerinaInt(Object a) {
    return a instanceof Number && !(a instanceof Double);
}
```

Reply:
@SasinduDilshara  Can you provide the scenario when run time value of int is not Long?

Reply:
Updated [aef3dc8](https://github.com/ballerina-platform/module-ballerina-data.jsondata/pull/2/commits/aef3dc8abdd5f918b9b14ae1036f935be8739e64)